docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+
+Void Type
+---------
+
+The type `void` contains no data and has a single, unnamed, value.
+
+A `void` value takes up no space, and thus does not affect the layout of types.
+Formally, a `void` value behaves as if it has a size of zero bytes, and one-byte alignment.
+
+Scalar Types
+------------
+
+### Boolean Type
+
+The type `bool` is used to represent Boolean truth value: `true` and `false`.
+
+The size of a `bool` varies across target platforms; programs that need to ensure a matching in-memory layout between targets should not use `bool` for in-memory data structures.
+On all platforms, the `bool` type must be _naturally aligned_ (its alignment is its size).
+
+### Integer Types
+
+The following integer types are defines:
+
+| Name          | Description |
+|---------------|-------------|
+| `int8_t`      | 8-bit signed integer |
+| `int16_t`     | 16-bit signed integer |
+| `int`         | 32-bit signed integer |
+| `int64_t`     | 64-bit signed integer |
+| `uint8_t`     | 8-bit unsigned integer |
+| `uint16_t`    | 16-bit unsigned integer |
+| `uint`        | 32-bit unsigned integer |
+| `uint64_t`    | 64-bit unsigned integer |
+
+All signed integers used two's complement representation.
+All arithmetic operations on integers (both signed and unsigned) wrap on overflow/underflow.
+
+All target platforms must support the `int` and `uint` types.
+Specific target platforms may not support the other integer types.
+
+All integer types are stored in memory with their natural size and alignment on all targes that support them.
+
+### Floating-Point Types
+
+The following floating-point type are defined:
+
+| Name          | Description                   |
+|---------------|-------------------------------|
+| `half`        | 16-bit floating-point number (1 sign bit, 5 exponent bits, 10 fraction bits) |
+| `float`       | 32-bit floating-point number (1 sign bit, 8 exponent bits, 23 fraction bits) |
+| `double`      | 64-bit floating-point number (1 sign bit, 11 exponent bits, 52 fraction bits) |
+
+All floating-point types are laid out in memory using the matching IEEE 754 standard format (`binary16`, `binary32`, `binary64`).
+Target platforms may define their own rules for rounding, precision, denormals, infinities, and not-a-number values.
+
+All target platforms must support the `float` type.
+Specific targets may not support the other floating-point types.
+
+All floating-point types are stored in memory with their natural size and alignment on all targets that support them.
+
+Vector Types
+------------
+
+A vector type is written as `vector<T, N>` and represents an `N`-element vector with elements of type `T`.
+The _element type_ `T` must be one of the built-in scalar types, and the _element count_ `N` must be a specialization-time constant integer.
+The element count must be between 2 and 4, inclusive.
```

Comment:
Whether or not to allow 1-element vectors, or matrices with only 1 row or column is something we can fine tune over time.

There is definitely a practical issue that GLSL doesn't allow 1-element vectors or matrices with only 1 row/column. Thus if we allow these in the front-end we'd have to include a pass to lower single-element vectors into scalars (and have a similar treatment for matrices, although ideally we will desugar matrices into arrays and avoid the whole problem).

In terms of the Slang front-end, the immediate question is: where would one-element vectors ever come from?

We could try to simplify our handling of swizzles so that a swizzle with `M` components always yields a `vector<M,T>` instead of having special-case rules for scalars. That sounds like a nice generalization (and mirrors your point about generic code being easier to write if you support one-element vectors), but it has the problem that if somebody writes:

```hlsl
float f = float3(99).z;
```

we need to have an explanation for how that code type-checks, which means we need implicit conversions from the hypothetical `float1` type to/from `float`, and those conversions need to be extremely low cost to make sure that things like overload resolution treat `float1` as basically the same as `float`.

We can implement all of that extra stuff (and that is what Cg and the "real" HLSL language do), but it has only a marginal impact on users (sometimes they will see `float1` instead of `float` in overload resolution errors, and wonder where it came from).

You also end up needing careful rules so that if, say, a user adds a `float4` and a `float1` then we *don't* handle that as a case of mismatched dimensions (like we would for a `float4` plus a `float3`) and instead handle it by casting from `float1 -> float -> float4` which is a bit weird and needs to be special-cased in the guts of the compiler (since things like generic argument inference already need to understand the scalar->vector case).

In the long run, I think I agree with your sentiment, though. We should realistically support vectors with any (non-negative) number of elements, and then legalize these based on what a given target can support. The special case of one-element vectors is just interesting because we probably have to allow them to easily convert to/from scalars in all contexts, which makes one-element vectors a big special case.